sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaCoder.java

```diff
@@ -100,4 +99,47 @@ public boolean consistentWithEquals() {
   public String toString() {
     return "SchemaCoder: " + rowCoder.toString();
   }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    SchemaCoder<?> that = (SchemaCoder<?>) o;
+    return rowCoder.equals(that.rowCoder)
+        && toRowFunction.equals(that.toRowFunction)
+        && fromRowFunction.equals(that.fromRowFunction);
```

Comment:
Doesn't this just revert to object equality comparison on the to/from functions?

Reply:
Yeah - I discussed this offline a bit with @kennknowles and he convinced me that it was better to have an equals function that might have some false negatives (if the toRowFunction and fromRowFunction don't have a good equals), rather than one that could have false positives (like if we rely on just checking the schema and typeDescriptor, and assume that the toRow/fromRow are the same).

I managed to make the CloudObjectsTest work by adding RowIdentity with an equals() function here.

Reply:
The way I would phrase this is: let the functions own their equals. If they say they are equal, they are. If they say they aren't, they aren't. So this equals() is relative to that.

Reply:
Could we go ahead and merge this as is? I could follow up with more changes to the SchemaCoder equals (plumbing through a type descriptor and using that for comparison, as well as possibly changing the toRow/fromRow functions created by the existing SchemaProviders to make them comparable)

Reply:
I have a PR up now (https://github.com/apache/beam/pull/9493) that adds `equals` and `hashCode` to the `fromRow` and `toRow` functions created by all the `GetterBasedSchemaProvider` sub-classes.

Reply:
BTW this is not just for tests. The Flink runner appears to rely on coder equality (even though you can argue it shouldn't).